### ISSUE TITLE
Rails5 beta3 url_for appends unwanted parameters

### ISSUE THREAD

**Issue by mfazekas** [rails/rails#24438]
### Steps to reproduce

```ruby
class ItemsController
  def update
    redirect_to({action: :index})
  end
  def index
    render plain: 'index'
  end
end
...
  def test_redirect
    patch(:update,{id:12})
    assert_redirected_to({action: :index})
    # fails with:
    # Expected response to be a redirect to <http://test.host/items?id=12> but was a redirect to <http://test.host/items>.

  end
```
see https://gist.github.com/mfazekas/922719bf65b3daeb7a0360e0ae7e7ebc for full repro script

was fine in rails 4.2

### Expected behavior
test should pass

### Actual behavior
id seems to be added to ```assert_redirected_to :index```

```
  1) Failure:
ItemsControllerTest#test_redirect [test_rails_redirect.rb:61]:
Expected response to be a redirect to <http://test.host/items?id=12> but was a redirect to <http://test.host/items>.
Expected "http://test.host/items?id=12" to be === "http://test.host/items".
```

### System configuration
**Rails version**:5.0.0.beta3
**Ruby version**:2.3.0

**Comment by mfazekas** [comment]
Reverting https://github.com/rails/rails/commit/1993e2ccbd7c5651278ea30bdc9d8034f5197945 fixes it.
I assume that not cloning the hashes and reusing hashes in other part of code causes this strange effect.

**Comment by maclover7** [comment]
... there isn't an index action in your test script's controller? Try to add and then run again.

**Comment by mfazekas** [comment]
@maclover7 i've added index but as expected it doesn't makes any difference, as index is newer called just redirected to.

**Comment by mfazekas** [comment]
I've added another testcase that shows that the issue is in url_for:
```ruby
  def edit
    render plain: "#{url_for(action: :index)} #{url_for(action: :index)}"
  end
```
i'd expect the two url_for to return the same thing but the second has the id appended
```ruby
Expected: "http://test.host/items"
  Actual: "http://test.host/items?id=12"
```

**Comment by vipulnsward** [comment]
Tried investigating this. We are mutating  `@recall` here https://github.com/rails/rails/blob/eba0041c2132e46bf9e040b31059e020e590d2ae/actionpack/lib/action_dispatch/routing/route_set.rb#L636

which is causing conditions on https://github.com/rails/rails/blob/eba0041c2132e46bf9e040b31059e020e590d2ae/actionpack/lib/action_dispatch/routing/route_set.rb#L568 to vary and result into different routes.

**Comment by vipulnsward** [comment]
Ahh, I see its being worked on here- https://github.com/rails/rails/issues/24392
 and https://github.com/rails/rails/pull/23103

**Comment by maclover7** [comment]
To recap open PRs for this issue -- #23103 and #24450 are open to solve this issue.

**Comment by jeremy** [comment]
Fix in #23103
